### PR TITLE
Refactor gatekeeper components

### PR DESCRIPTION
**Pull request by p-netm**
- Added an prop enum that can be used to configure the authorization
grant type i.e. between implicit and autorization code.
- Add headless login hook to login component
- refactoring magic strings & break down login tests

### CODE REVIEW

File: packages/gatekeeper/src/components/login.tsx [hunk 1]
```diff
@@ -1,34 +1,71 @@
 import React from 'react';
+import { OauthLoginProps } from '..';
```
Comment (moshthepitt) [inline packages/gatekeeper/src/components/login.tsx]: I think this is defined within this module

File: packages/gatekeeper/src/components/login.tsx [hunk 2]
```diff
@@ -1,34 +1,71 @@
...
-  providers: Providers;
+export interface ProviderLinksProps extends OAuthLoginHookOptions {
+  OAUTH_LOGIN_PROMPT?: string;
+}
+
+/** interface for OauthLogin props */
+export interface OauthLoginProps extends ProviderLinksProps {
+  ProviderLinksComponent?: React.ElementType;
+  NO_PROVIDERS?: string;
+}
```
Comment (moshthepitt) [inline packages/gatekeeper/src/components/login.tsx]: I think you should not mark these properties as optional.  I think of it this way: 

- can this property make any sense with the optional properties excluded
- is it justifiable to "weaken" the type definition like this?
- for React components, Typescript works with "defaultProps" to negate the need to mark anything as optional e.g. here https://github.com/onaio/js-tools/blob/master/packages/gatekeeper/src/components/callback.tsx#L29

File: packages/gatekeeper/src/components/login.tsx [hunk 3]
```diff
@@ -1,34 +1,71 @@
...
-export interface ProviderLinksProps {
-  providers: Providers;
+export interface ProviderLinksProps extends OAuthLoginHookOptions {
+  OAUTH_LOGIN_PROMPT?: string;
+}
+
+/** interface for OauthLogin props */
+export interface OauthLoginProps extends ProviderLinksProps {
+  ProviderLinksComponent?: React.ElementType;
+  NO_PROVIDERS?: string;
```
Comment (moshthepitt) [inline packages/gatekeeper/src/components/login.tsx]: I think we should maintain normal/regular case rules for object property names (even if it refers to something that is populated from a constant)

File: packages/gatekeeper/src/components/login.tsx [hunk 4]
```diff
@@ -1,34 +1,71 @@
...
+export interface OAuthLoginHookOptions {
   providers: Providers;
+  authorizationGrant: AuthorizationGrantType;
 }
 
 /** interface for ProviderLinks props */
-export interface ProviderLinksProps {
-  providers: Providers;
+export interface ProviderLinksProps extends OAuthLoginHookOptions {
+  OAUTH_LOGIN_PROMPT?: string;
```
Comment (moshthepitt) [inline packages/gatekeeper/src/components/login.tsx]: I think we should maintain normal/regular case rules for object property names (even if it refers to something that is populated from a constant)

File: packages/gatekeeper/src/components/tests/login.test.tsx [hunk 5]
```diff
@@ -21,7 +21,9 @@ describe('gatekeeper/OauthLogin', () => {
       providers: fixtures.providers
```
Comment (moshthepitt) [inline packages/gatekeeper/src/components/tests/login.test.tsx]: Lets add a test that uses both types of OAuth flows to codify what the expected URL should be

File: packages/gatekeeper/src/components/login.tsx (hunk 2 above)
Comment (p-netm) [reply packages/gatekeeper/src/components/login.tsx]: Made them required.

File: packages/gatekeeper/src/components/login.tsx (hunk 3 above)
Comment (p-netm) [reply packages/gatekeeper/src/components/login.tsx]: ok.

File: packages/gatekeeper/src/components/tests/login.test.tsx (hunk 5 above)
Comment (p-netm) [reply packages/gatekeeper/src/components/tests/login.test.tsx]: on it

File: packages/gatekeeper/README.md [hunk 6]
```diff
@@ -96,7 +97,7 @@ The logout component takes these props that are useful in extending it:
 
 ### oAuth2 Login Component
 
-GateKeeper provides a simple login page component to help start the oAuth2 process.
+GateKeeper provides a simple login page component and hook to help start the oAuth2 process.
```
Comment (moshthepitt) [inline packages/gatekeeper/README.md]: > start the oAuth2 process

What do you mean?

**Pull request merged**